torchdata/nodes/samplers/multi_node_weighted_sampler.py

```diff
@@ -0,0 +1,216 @@
+                f"Invalid {self.stop_criteria=}. stop_criteria must be one of: CYCLE_UNTIL_ALL_DATASETS_EXHAUSTED, FIRST_DATASET_EXHAUSTED, ALL_DATASETS_EXHAUSTED"
+            )
+
+        # Check if keys of source_nodes and weights are the same
+        if set(self.dataset_names) != set(self.weights.keys()):
+            raise ValueError(
+                f"Invalid {self.weights=}. For multi-dataset weighted sampling, keys of source_nodes and weights must be the same",
+            )
+
+        for weight in self.weights.values():
+            if not isinstance(weight, float) or weight <= 0:
+                raise ValueError(
+                    f"Invalid {self.weights=}. For multi-dataset weighted sampling, weights must be a 1d sequence, non-negative, and non-zero"
+                )
+
+    def reset(self, initial_state: Optional[Dict[str, Any]] = None):
+        super().reset(initial_state)
+
+        if initial_state is not None:
+            self._num_yielded = initial_state[self.NUM_YIELDED_KEY]
+            self._weighted_sampler = _WeightedSampler(
+                weights=self.weights,
+                seed=self.seed,
+                rank=self.rank,
+                world_size=self.world_size,
+                initial_state=initial_state[self.WEIGHTED_SAMPLER_STATE_KEY],
+            )
+            self._datasets_exhausted = initial_state[self.DATASETS_EXHAUSTED_KEY]
+            for k in self.dataset_names:
+                self.source_nodes[k].reset(initial_state[self.DATASET_NODE_STATES_KEY][k])
+        else:
+            # Force a fresh iterator from all source nodes
+            self._num_yielded = 0
+            self._weighted_sampler = _WeightedSampler(
+                weights=self.weights,
+                seed=self.seed,
+                rank=self.rank,
+                world_size=self.world_size,
+            )
+            self._datasets_exhausted = {key: False for key in self.weights.keys()}
+            for k in self.dataset_names:
+                self.source_nodes[k].reset()
+
+    def next(self) -> T:
+        if all(self._datasets_exhausted.values()):
+            # Raise StopIteration if all datasets are exhausted
+            raise StopIteration()
+
+        # Raise StopIteration is StopCriteria is FIRST_DATASET_EXHAUSTED and
+        # the first dataset is exhausted. Doing this to correctly catch StopIteration
+        # when trying next(it) on already exhausted iterator
+        if self.stop_criteria == StopCriteria.FIRST_DATASET_EXHAUSTED and any(self._datasets_exhausted.values()):
+            raise StopIteration()
+
+        key = next(self._weighted_sampler)
+        try:
+            if self._datasets_exhausted[key] and self.stop_criteria == StopCriteria.ALL_DATASETS_EXHAUSTED:
+                # Before fetching a new item check if key corresponds to an already
+                # exhaused dataset and StopCriteria is ALL_DATASETS_EXHAUSTED, move to next idx
+                return self.next()
```

Comment:
I am pretty sure this should be 
```suggestion
                return next(self)
```

Reply:
actually, can we avoid the recursion here? It becomes hard to reason about. eg if this throws StopIteration, would it get caught again in the try/except?